### PR TITLE
Improve chainer examples

### PR DESCRIPTION
**Pull request by crcrpar**
This PR partially accesses #6398 and related to #3500 and #6378.

- Avoid comparisons with empty string by setting `args.resume` as `None` by default.
- Extend `Trainer` instance with `extensions.snapshot`
- [custom loop] Fix ignorance of `args.resume` and add messages to be more user-friendly.

### CODE REVIEW

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
It seems the word2vec example failed to serialize the iterator.
```
+ coverage run -a --branch examples/word2vec/train_word2vec.py -e 1 -b 10 --test
GPU: -1
# unit: 100
Window: 5
Minibatch-size: 10
# epoch: 1
Training model: skipgram
Output type: hsm

n_vocab: 10000
data length: 100
epoch       main/loss   validation/main/loss
1           382.919     309.452               
Exception in main training loop: 'WindowIterator' object has no attribute '_order'
Traceback (most recent call last):
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 319, in run
    entry.extension(self)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/extensions/_snapshot.py", line 214, in __call__
    self._make_snapshot(trainer)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/extensions/_snapshot.py", line 218, in _make_snapshot
    serialized_target = npz.serialize(target)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/serializers/npz.py", line 67, in serialize
    s.save(obj)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/serializer.py", line 67, in save
    obj.serialize(self)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 361, in serialize
    self.updater.serialize(serializer['updater'])
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/updaters/standard_updater.py", line 194, in serialize
    iterator.serialize(serializer['iterator:' + name])
  File "examples/word2vec/train_word2vec.py", line 140, in serialize
    if self._order is not None:
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "examples/word2vec/train_word2vec.py", line 282, in <module>
    main()
  File "examples/word2vec/train_word2vec.py", line 270, in main
    trainer.run()
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 319, in run
    entry.extension(self)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/extensions/_snapshot.py", line 214, in __call__
    self._make_snapshot(trainer)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/extensions/_snapshot.py", line 218, in _make_snapshot
    serialized_target = npz.serialize(target)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/serializers/npz.py", line 67, in serialize
    s.save(obj)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/serializer.py", line 67, in save
    obj.serialize(self)
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 361, in serialize
    self.updater.serialize(serializer['updater'])
  File "/home/user/.local/lib/python2.7/site-packages/chainer/training/updaters/standard_updater.py", line 194, in serialize
    iterator.serialize(serializer['iterator:' + name])
  File "examples/word2vec/train_word2vec.py", line 140, in serialize
    if self._order is not None:
AttributeError: 'WindowIterator' object has no attribute '_order'
```

**Comment by crcrpar** [comment]
uh-oh...

**Comment by crcrpar** [comment]
Fixed (it was a typo...) and I confirmed the example kicked by `python train_word2vec.py --test` finished without the failure.

**Comment by toslunar** [comment]
The bug probably came from that the draft implementation used `SerialIterator`, which had `_order` (before #5669).

**Comment by toslunar** [comment]
Jenkins, test this please.

File: examples/word2vec/train_word2vec.py [hunk 1]
```diff
@@ -251,10 +260,17 @@ def main():
...
+        extensions.snapshot(filename='snapshot_epoch_{.updater.epoch}'),
+        trigger=(args.snapshot_interval, 'epoch'))
+
+    if args.resume is not None:
+        chainer.serializers.load_npz(args.resume, trainer)
     trainer.run()
 
     # Save the word2vec model
-    with open('word2vec.model', 'w') as f:
+    with open(os.path.join(args.out, 'word2vec.model'), 'w') as f:
```
Comment (toslunar) [inline examples/word2vec/train_word2vec.py]: The inference example will not use the trained model.
https://github.com/chainer/chainer/blob/645483f9fd994f6b7dfa39f4d0c098a7177be0b3/examples/word2vec/search.py#L8

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**